vscode-antimony/src/server/stibium/stibium/analysis.py

```diff
@@ -429,6 +440,37 @@ def handle_annotation(self, scope: AbstractScope, annotation: Annotation):
         qname = QName(scope, name)
         self.table.insert(qname, SymbolType.Parameter)
         self.table.insert_annotation(qname, annotation)
+
+    def pre_handle_rate_rule(self, scope, rate_rule):
+        self.pending_rate_rules.append((scope, rate_rule))
+
+    def handle_rate_rules(self):
+        for scope, rate_rule in self.pending_rate_rules:
+            self.handle_rate_rule(scope, rate_rule)
+
+    def handle_rate_rule(self, scope, rate_rule : RateRules):
+        name = rate_rule.get_name()
+        qname = QName(scope, name)
+        expression = rate_rule.get_value()
+        if len(self.table.get(qname)) != 0:
+            var = self.table.get(qname)[0]
+            if var.type == SymbolType.Species and var.in_reaction and not var.is_const:
+                self.error.append(RateRuleNotInReaction(rate_rule.range, name.text))
+            rate_rule_string = ""
+            for leaf in expression.scan_leaves():
+                if isinstance(leaf, Name) and leaf.text not in self.table.get_all_names():

```

Comment:
move this outside of the for loop for better performance. Otherwise, the names get looped over, and the set gets created every time.